vpn/tls.go

```diff
@@ -28,53 +28,144 @@ var (
 	ErrBadKeypair = errors.New("bad keypair conf")
 	// ErrBadParrot is returned for errors during TLS parroting
 	ErrBadParrot = errors.New("cannot parrot")
+	// ErrCannotVerifyCertChain is returned for certificate chain validation errors.
+	ErrCannotVerifyCertChain = errors.New("cannot verify chain")
 )
 
-// initTLS returns a tls.Config matching the VPN options.
-func initTLS(session *session, opt *Options) (*tls.Config, error) {
-	if session == nil || opt == nil {
+// certVerifyOptionsNoCommonNameCheck returns a x509.VerifyOptions initialized with
+// an empty string for the DNSName. This allows to skip CN verification.
+func certVerifyOptionsNoCommonNameCheck() x509.VerifyOptions {
+	return x509.VerifyOptions{DNSName: ""}
+}
+
+// certVerifyOptions is the options fatory that the customVerify function will
+// use; by default it configures VerifyOptions to skip the DNSName check.
+var certVerifyOptions = certVerifyOptionsNoCommonNameCheck
+
+// certPaths holds the paths for the cert, key, and ca used for OpenVPN
+// certificate authentication.
+type certPaths struct {
+	certPath string
+	keyPath  string
+	caPath   string
+}
+
+// loadCertAndCA parses the PEM certificates contained in the paths pointed by
+// certAuthConfig and return a certAuth with the client and CA certificates.
+func loadCertAndCA(pth certPaths) (*certConfig, error) {
+	ca := x509.NewCertPool()
+	caData, err := ioutil.ReadFile(pth.caPath)
+	if err != nil {
+		return nil, fmt.Errorf("%w:%s", ErrBadCA, err)
+	}
+	ok := ca.AppendCertsFromPEM(caData)
+	if !ok {
+		return nil, fmt.Errorf("%w:%s", ErrBadCA, "cannot parse ca cert")
+	}
+
+	cert, err := tls.LoadX509KeyPair(pth.certPath, pth.keyPath)
+	if err != nil {
+		return nil, fmt.Errorf("%w:%s", ErrBadKeypair, err)
+	}
+	cfg := &certConfig{
+		ca:   ca,
+		cert: cert,
+	}
+	return cfg, nil
+}
+
+// authorityPinner is any object from which we can obtain a certpool containing
+// a pinned Certificate Authority for verification.
+type authorityPinner interface {
+	authority() *x509.CertPool
+}
+
+// certConfig holds the parsed certificate and CA used for OpenVPN mutual
+// certificate authentication.
+type certConfig struct {
+	cert tls.Certificate
+	ca   *x509.CertPool
+}
+
+// newCertConfigFromOptions is a constructor that returns a certConfig object initialized
+// from the paths specified in the passed Options object, and an error if it
+// could not be properly built.
+func newCertConfigFromOptions(o *Options) (*certConfig, error) {
+	return loadCertAndCA(certPaths{
+		certPath: o.Cert,
+		keyPath:  o.Key,
+		caPath:   o.Ca,
+	})
+}
+
+// authority implements authorityPinner interface.
+func (c *certConfig) authority() *x509.CertPool {
+	return c.ca
+}
+
+// ensure certConfig implements authorityPinner.
+var _ authorityPinner = &certConfig{}
+
+// verifyFun is the type expected by the VerifyPeerCertificate callback in tls.Config.
+type verifyFun = func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error
```

Comment:
```suggestion
type verifyFun func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error
```

No need to use type alias